src/validationinterface.cpp

```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
     boost::signals2::signal<void (const CBlockIndex *, const std::shared_ptr<const CBlock>&)> NewPoWValidBlock;
 
     CScheduler *scheduler = NULL;
+
+    // We are not allowed to assume the scheduler only runs in one thread,
+    // but must ensure all callbacks happen in-order, so we end up creating
+    // our own queue here :(
+    CCriticalSection cs_callbacksPending;
+    std::list<std::function<void (void)>> callbacksPending;
+    bool fCallbacksRunning = false;
+
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
```

Comment:
@mchrostowski hmm, really, I find that it decreases readability (though that may be NIH). It looks harder to reason about whether some callbacks might accidentally get missed to me.

(Other random note, we dont use tabs in our codebase, which your patch added).